pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,244 @@
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
+	}
+
+	path := lbls[PrepDownscalePathKey]
+	if path == "" {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePathKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePathKey))
+		return &reviewResponse
+	}
+
+	rolloutGroup := lbls[RolloutGroupLabelKey]
+	if rolloutGroup != "" {
+		foundSts, err := findDownscalesDoneMinTimeAgo(ctx, api, ar.Request.Namespace, ar.Request.Name, rolloutGroup)
+		if err != nil {
+			reviewResponse := v1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because finding other statefulsets failed.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas),
+				},
+			}
+			level.Warn(logger).Log("msg", "downscale not allowed due to error while adding annotation", "err", err)
+			return &reviewResponse
+		}
+		if foundSts != nil {
+			reviewResponse := v1.AdmissionResponse{
+				Allowed: false,
+				Result: &metav1.Status{
+					Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because statefulset %v was downscaled at %v and is labelled to wait %s between zone downscales", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, foundSts.name, foundSts.lastDownscaleTime, foundSts.waitTime),
+				},
+			}
+			level.Warn(logger).Log("msg", "downscale not allowed because another statefulset was downscaled recently", "err", err)
+			return &reviewResponse
+		}
+	}
+
+	if *ar.Request.DryRun {
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it needs to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	diff := (*oldReplicas - *newReplicas)
+	type endpoint struct {
+		url   string
+		index int
+	}
+	eps := make([]endpoint, diff)
+
+	// The DNS entry for a pod of a stateful set is
+	// ingester-zone-a-0.$(servicename).$(namespace).svc.cluster.local
+	// The service in this case is ingester-zone-a as well.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+	for i := 0; i < int(diff); i++ {
+		index := int(*oldReplicas) - i - 1 // nr in statefulset
+		eps[i].url = fmt.Sprintf("%v-%v.%v.%v.svc.cluster.local:%s/ingester/%s",
```

Comment:
Can `/ingester/` be a part of the path rather than being hardcoded?

Reply:
Yes, that can be removed.

Reply:
@andyasp I just found out this will not work. A label cannot contain `/`. So I'll have to reinstate that `/ingester/` prefix for now.